src/dune_engine/targets.mli

```diff
@@ -63,3 +63,37 @@ val head : t -> Path.Build.t option
 val to_dyn : t -> Dyn.t
 
 val pp : t -> _ Pp.t
+
+(** The set of targets produced by an action. Each target may be tagged with a
+    payload, for example, the target's digest. *)
+module Produced : sig
+  type 'a t = private
+    { files : 'a Path.Build.Map.t
+    ; dirs : 'a Path.Build.Map.t Path.Build.Map.t
+    }
+
+  (** Returns the given [targets : Validated.t]. Raises a code error if
+      [on_dir_target = `Raise] and [targets.dir] is non-empty. *)
+  val of_validated_files :
+    Validated.t -> on_dir_target:[< `Ignore | `Raise ] -> unit t
+
+  (** Populates only the [files] field, leaving [dirs] empty. Raises a code
+      error if the list contains duplicates. *)
+  val of_file_list_exn : (Path.Build.t * Digest.t) list -> Digest.t t
+
+  (** Add a list of discovered directory-file pairs to [Validated.t]. Raises a
+      code error on an unexpected directory. *)
+  val expand_validated_exn :
+    Validated.t -> (Path.Build.t * Path.Build.t) list -> unit t
+
+  (** Union of [t.files] and all files in [t.dirs]. *)
+  val all_files : 'a t -> 'a Path.Build.Map.t
+
+  (** Aggregate all content digests. *)
+  val digest : Digest.t t -> Digest.t
+
+  val with_digests :
```

Comment:
Any reason this is specialised to digests? It seems that we could replace `Digest.t` by `'a` and it would work just as well. It feels that this is a special form of `map`. Maybe it should be called `try_map`?